sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -318,17 +318,31 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
 
   /**
    * Statically evaluate an expression containing zero or more placeholders, given a set
-   * of bindings for placeholder values.
+   * of bindings for placeholder values, if the expression is evaluable. If it is not,
+   * bind statically evaluated expression results to an expression.
    */
-  private def evalExpr(expr: Expression, bindings: Map[ExprId, Option[Any]]) : Option[Any] = {
+  private def bindingExpr(
+      expr: Expression,
+      bindings: Map[ExprId, Option[Expression]]): Option[Expression] = {
     val rewrittenExpr = expr transform {
       case r: AttributeReference =>
         bindings(r.exprId) match {
-          case Some(v) => Literal.create(v, r.dataType)
+          case Some(v) => v
           case None => Literal.default(NullType)
         }
     }
-    Option(rewrittenExpr.eval())
+    if (!rewrittenExpr.foldable) {
```

Comment:
shall we apply the check in more places? `evalAggOnZeroTups` also calls `eval()` directly.

Reply:
yes. this is not possible for PythonUDF, but it is potential for other not foldable expression.

Reply:
so it is not covered by added test. Let me add test for it...